File: scripts/verify/outer_facade_parity.py
# -*- coding: utf-8 -*-
"""Compare the Python tower bucketing against what the browser actually did.

Reads scripts/verify/out/outer-facade-browser.json, written by
scripts/verify/outer-facade-parity.mjs from a live run of
`window.quantiseOuterFacades`.

THE TEST IS THE PARTITION. The two sides cannot agree on the bucket NAME — the
browser's `tg<n>` counts from the end of the campus palette, which is the whole
reason the assignment is being moved into the bake — so what is checked is:

  1. every tower the browser stamped, Python also stamps;
  2. two towers share a bucket in Python if and only if they share one in the
     browser (a bijection between the two labellings, checked both ways, so a
     Python bucket that merges two browser buckets fails and so does the split);
  3. the centroid colours agree to within a rounding step.

Point 2 both ways is the part worth having. A one-way check passes happily when
Python collapses ten buckets into three.

Usage:  python scripts/verify/outer_facade_parity.py
"""
import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(os.path.dirname(HERE))
sys.path.insert(0, os.path.join(ROOT, "scripts"))

from bake_outer_facades import (  # noqa: E402
    cluster_colours, bucket_of, hex_to_rgb, to_hex, TOWER_BUCKETS)

BROWSER = os.path.join(HERE, "out", "outer-facade-browser.json")


def main():
    if not os.path.exists(BROWSER):
        print("*FAIL — %s is missing. Run outer-facade-parity.mjs first."
              % os.path.relpath(BROWSER, ROOT))
        return 1
    b = json.load(open(BROWSER, encoding="utf-8"))
    wd, browser_wp = b["wd"], b["wp"]
    if len(wd) != len(browser_wp):
        print("*FAIL — malformed capture")
        return 1

    cent = cluster_colours(wd, TOWER_BUCKETS)
    py_wp = ["tb%02d" % bucket_of(h, cent) for h in wd]

    fails = []
    if any(x is None for x in browser_wp):
        fails.append("browser left %d towers unstamped"
                     % sum(1 for x in browser_wp if x is None))

    # A bijection between the two labellings, checked in both directions.
    fwd, rev = {}, {}
    for i, (p, q) in enumerate(zip(py_wp, browser_wp)):
        if q is None:
            continue
        if fwd.setdefault(p, q) != q:
            fails.append("python %s maps to both %s and %s (tower %d)"
                         % (p, fwd[p], q, i))
        if rev.setdefault(q, p) != p:
            fails.append("browser %s maps to both %s and %s (tower %d)"
                         % (q, rev[q], p, i))

    # Centroids: group means on each side, matched through the bijection.
    def means(labels):
        acc = {}
        for lab, h in zip(labels, wd):
            r, g, bl = hex_to_rgb(h)
            a = acc.setdefault(lab, [0, 0, 0, 0])
            a[0] += r; a[1] += g; a[2] += bl; a[3] += 1
        return {k: to_hex([v[0] / v[3], v[1] / v[3], v[2] / v[3]])
                for k, v in acc.items()}

    pm, bm = means(py_wp), means(browser_wp)
    for p, q in fwd.items():
        if pm[p] != bm[q]:
            fails.append("centroid mismatch %s %s vs %s %s" % (p, pm[p], q, bm[q]))

    sizes = {}
    for p in py_wp:
        sizes[p] = sizes.get(p, 0) + 1

    print("towers                 %d" % len(wd))
    print("python buckets         %d" % len(sizes))
    print("browser buckets        %d" % len(rev))
    print("sizes                  " + "  ".join(
        "%s=%d" % (k, sizes[k]) for k in sorted(sizes)))
    print("bucket map             " + "  ".join(
        "%s->%s" % (k, fwd[k]) for k in sorted(fwd)))
    print("")
    if fails:
        print("*FAIL — %d finding(s)" % len(fails))
        for f in fails[:20]:
            print("   " + f)
        return 1
    print("PASS — the bake partitions the towers exactly as the browser does, "
          "and the group centroids agree.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/verify/outer-facade-parity.mjs
/**
 * outer-facade-parity.mjs — does the Python port bucket the towers the same way
 * the browser does?
 *
 * WHAT IS BEING PROVED, precisely. `scripts/bake_outer_facades.py` is a
 * transcription of `clusterColours` and the tower loop in `js/facades.js`. A
 * transcription checked by re-reading it proves nothing — the same
 * misunderstanding writes both sides. So this runs the REAL browser function on
 * the REAL data and dumps what it decided, and the Python side is compared
 * against that.
 *
 * IT COMPARES THE PARTITION, NOT THE IDS, and that is not a dodge. The browser
 * names a bucket `tg<n>` where n is an offset into the campus palette, so the
 * two sides cannot agree on the string by construction — that offset is the
 * whole reason the port exists. What has to match is which towers land
 * together, and the centroid each group sits on.
 *
 * Note it calls `quantiseOuterFacades` on features it fetched itself rather
 * than on js/outer.js's copy. That is deliberate: it keeps the check
 * independent of whether the ring happens to be on tiles or on GeoJSON today,
 * and on the tile path js/outer.js never fetches the document at all.
 *
 * Writes scripts/verify/out/outer-facade-browser.json.
 * Compare with:  python scripts/verify/outer_facade_parity.py
 *
 * Usage: VERIFY_URL=http://127.0.0.1:8123 node scripts/verify/outer-facade-parity.mjs
 */
import { chromium } from 'playwright-core';
import { BASE, launch } from './chrome.mjs';
import fs from 'node:fs';
import path from 'node:path';

const OUT_DIR = path.join('scripts', 'verify', 'out');
fs.mkdirSync(OUT_DIR, { recursive: true });

const browser = await launch(chromium);
const page = await browser.newPage({ viewport: { width: 900, height: 600 } });
page.on('pageerror', e => console.log('  [pageerror] ' + e.message));

await page.goto(BASE + '/index.html?intro=0&drift=0', { waitUntil: 'networkidle', timeout: 120000 });
await page.waitForFunction(() => window.__map && window.__map.isStyleLoaded(), null, { timeout: 120000 });
await page.evaluate(() => window.cancelGraphicsAutoDetect && window.cancelGraphicsAutoDetect());
await page.evaluate(() => new Promise(r => {
  const m = window.__map;
  if (m.loaded()) return r();
  m.once('idle', r); setTimeout(r, 60000);
}));

const result = await page.evaluate(async () => {
  if (typeof window.quantiseOuterFacades !== 'function') {
    return { error: 'quantiseOuterFacades is not defined' };
  }
  const gj = await fetch('data/outer_ring.geojson').then(r => r.json());
  // The same predicate the browser pass uses.
  const towers = gj.features.filter(
    f => f.properties && f.properties.t === 1 && f.properties.wd);
  const n = window.quantiseOuterFacades(towers, window.__map);
  return {
    stamped: n,
    towers: towers.length,
    // File order is the join key: both sides read the same document and neither
    // reorders it. There is no id in the data to key on — checked.
    wp: towers.map(f => f.properties.wp || null),
    wd: towers.map(f => f.properties.wd),
  };
});

if (result.error) {
  console.log('*FAIL — ' + result.error);
  process.exitCode = 1;
} else {
  const groups = new Map();
  for (const id of result.wp) if (id) groups.set(id, (groups.get(id) || 0) + 1);
  console.log(`browser stamped ${result.stamped} of ${result.towers} towers`);
  console.log(`distinct buckets: ${groups.size}`);
  console.log('sizes: ' + [...groups.entries()]
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([k, v]) => `${k}=${v}`).join('  '));
  const file = path.join(OUT_DIR, 'outer-facade-browser.json');
  fs.writeFileSync(file, JSON.stringify(result, null, 1));
  console.log('\nwrote ' + file);
  console.log('now run:  python scripts/verify/outer_facade_parity.py');
  if (result.wp.some(x => !x)) {
    console.log('*FAIL — some towers were left unstamped by the browser');
    process.exitCode = 1;
  }
}

await browser.__done();
